Salesforce/client/token_refresher.py

```diff
@@ -0,0 +1,160 @@
+
+
+class SalesforceTokenRefresher(object):
+    """
+    Contains access token refresher logic.
+
+    Example of usage:
+    >>> from client.token_refresher import SalesforceTokenRefresher
+    >>>
+    >>> token_refresher = SalesforceTokenRefresher(**config)
+    >>>
+    >>> async with token_refresher.with_access_token() as access_token:
+    >>>     print(access_token)
+    """
+
+    _instances: dict[str, "SalesforceTokenRefresher"] = {}
+    _locks: dict[str, Lock] = {}
+    _session: ClientSession | None = None
+
+    def __init__(self, client_id: str, client_secret: str, auth_url: str, token_ttl: int = 3600):
+        """
+        Initialize SalesforceTokenRefresher.
+
+        Each set of scopes will have its own access token with refreshing logic.
+
+        Args:
+            client_id: str
+            client_secret: str
+            auth_url: str
+            token_ttl: int
+        """
+        self.client_id = client_id
+        self.client_secret = client_secret
+        self.auth_url = auth_url
+        self.token_ttl = token_ttl
+
+        self._token: SalesforceToken | None = None
+        self._token_refresh_task: Optional[Task[None]] = None
+
+    @classmethod
+    def session(cls) -> ClientSession:
+        """
+        Initialize client session.
+
+        Singleton client session to work with token refresh logic.
+
+        Returns:
+            ClientSession:
+        """
+        if not cls._session:
+            cls._session = ClientSession()
+
+        return cls._session
+
+    @classmethod
+    async def instance(
+        cls,
+        client_id: str,
+        client_secret: str,
+        auth_url: str,
+    ) -> "SalesforceTokenRefresher":
+        """
+        Get singleton SalesforceTokenRefresher instance for specified set of scopes.
+
+        Args:
+            client_id: str
+            client_secret: str
+            auth_url: str
+
+        Returns:
+            SalesforceTokenRefresher:
+        """
+        refresher_unique_key = str(frozenset({client_id, client_secret}))
+        if not cls._locks.get(refresher_unique_key):
+            cls._locks[refresher_unique_key] = asyncio.Lock()
+
+        if not cls._instances.get(refresher_unique_key):
+            async with cls._locks[refresher_unique_key]:
+                if not cls._instances.get(refresher_unique_key):
+                    cls._instances[refresher_unique_key] = SalesforceTokenRefresher(
+                        client_id,
+                        client_secret,
+                        auth_url,
+                    )
+
+        return cls._instances[refresher_unique_key]
+
+    async def refresh_token(self) -> None:
+        """
+        Refresh token based on class configuration.
+
+        Also triggers token refresh task.
+        """
+        params = {
+            "grant_type": "client_credentials",
+        }
+
+        url = URL("{0}/services/oauth2/token".format(self.auth_url)).with_query(urlencode(params, encoding="utf-8"))
```

Comment:
Instead of the `format` method, you can also use [f-string](https://docs.python.org/3/tutorial/inputoutput.html#tut-f-strings)
```suggestion
        url = URL(f"{self.auth_url}/services/oauth2/token").with_query(urlencode(params, encoding="utf-8"))
```

Reply:
I tried to find some time ago what is the best way of doing it in python, and found interesting discussion here
https://github.com/wemake-services/wemake-python-styleguide/issues/32
What do you think? 
I would like to make it more/less standard, so if we decide about f-strings, I'll use it everywhere